### ISSUE TITLE
[QUESTION] Mem usage is the same for 2 gpus and 4 gpus.

### ISSUE THREAD

**Issue by ilml** [microsoft/DeepSpeed#4093]
# Description
Hi team, I’m trying the a deepspeed project on two gpus, and I see that the mem usage per gpu is about 13G, but using the estimate_zero3_model_states_mem_needs function estimator, they mem usage should be about 0.77G.
And futher, if I use 4 gpus, the mem usage is the same as 2 gpus.
I think if we use zero-3, the mem usage should decrese linearly right.
Does anyone know why this is the case? Thanks a lot!

Model: gpt-j-6b

mem estimate:
![image](https://github.com/microsoft/DeepSpeed/assets/18288209/aa36b05d-e9d1-4a08-92ea-382103390418)

Actual usage on 2 gpus:
![image](https://github.com/microsoft/DeepSpeed/assets/18288209/c8121fc2-99bb-4cfe-bdb8-7e97b8658a65)

On 4 gpus(batch size is the same as 2 gpus):

![image](https://github.com/microsoft/DeepSpeed/assets/18288209/17df43f6-1e6e-4140-8a51-4d1ada2b45ef)

Config I use:

{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },

    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "overlap_comm": true,
        "contiguous_gradients": true,
        "sub_group_size": 1e9,
        "reduce_bucket_size": "auto",
        "stage3_prefetch_bucket_size": "auto",
        "stage3_param_persistence_threshold": "auto",
        "stage3_max_live_parameters": 1e9,
        "stage3_max_reuse_distance": 1e9,
        "stage3_gather_16bit_weights_on_model_save": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}

**Comment by PicoCreator** [comment]
Not sure if this is whats happening to you, but atleast what i observe in my projects (im not a dev for deepspeed)

Unless you are clearing the cache between each step (you probably shouldn't), DS3 will gladly use the remaining vram space and fill it up with the some weights in other GPUs if possible - which im guessing to avoid cross GPU communication - which can be a bottleneck

**Comment by ilml** [comment]
> Not sure if this is whats happening to you, but atleast what i observe in my projects (im not a dev for deepspeed)
> 
> Unless you are clearing the cache between each step (you probably shouldn't), DS3 will gladly use the remaining vram space and fill it up with some weights from other GPUs if possible - which im guessing is to avoid cross GPU communication - which can be a bottleneck
> 
> However if you do test it with GPUs with less vram, you get to see if it fits with less vram

Thanks for the information! I got your idea, so you're saying that, if we didn't set a large batch size, and there is extra space after plain DS3 is initialized, then DS3 will automatically fill these space with weights to accelerate the communication?

**Comment by PicoCreator** [comment]
yup, if you are looking at a DS2 vs DS3 vram usage.
Assuming the model fit when in DS2 mode.

I dun think it has anything to do with batch size. Because what DS3 does is that it removes duplicate copies of the weights across the GPU's, but this forces the GPU's to read from one another instead for those weight values.

This is great for super large models that dun fit into a GPU. But is a slow process, hence when possible it will try to cache some of those weights, all the way up to effectively the same as DS2.

PS: this is my understanding from my observation, we might still want someone official to confirm

**Comment by tjruwase** [comment]
> Hi team, I’m trying the a deepspeed project on two gpus, and I see that the mem usage per gpu is about 13G, but using the estimate_zero3_model_states_mem_needs function estimator, they mem usage should be about 0.77G.

@ilml, thanks for this question. 

First, looking closely at the estimator output, you will see that 0.77G per GPU is for the case of offloading both parameters and optimizers. 

Secondly, these estimates are only for model and optimizer state memory consumption and does not account for activation and gradient memory consumption that would occur during training. And so, the above screenshots are not appropriate comparison points.  One way to get a more precise memory consumption is to call the [see_memory_usage()](https://github.com/microsoft/DeepSpeed/blob/ff7d5275f2aa916cb5f320e0d817154e96f9cdb6/deepspeed/runtime/utils.py#L793) utility after model and optimizer initialization (e.g., `deepspeed.intialize()`). This utility leverages torch.cuda memory management [routines](https://pytorch.org/docs/stable/cuda.html#cuda-memory-management-api).

**Comment by kisseternity** [comment]
Hello, I have a question for ZeRO3 memory usage too. When I'm training a 13B model with 8 machines of 8 * A100, the maximum seq length can be set to 25k. However, when the machines num come to 12, 25k seq length will lead to OOM. As you say, the activation seems to be a bottleneck here, but don't know why the seq length can't be the same.

**Comment by kisseternity** [comment]
> > Hi team, I’m trying the a deepspeed project on two gpus, and I see that the mem usage per gpu is about 13G, but using the estimate_zero3_model_states_mem_needs function estimator, they mem usage should be about 0.77G.
> 
> @ilml, thanks for this question.
> 
> First, looking closely at the estimator output, you will see that 0.77G per GPU is for the case of offloading both parameters and optimizers.
> 
> Secondly, these estimates are only for model and optimizer state memory consumption and does not account for activation and gradient memory consumption that would occur during training. And so, the above screenshots are not appropriate comparison points. One way to get a more precise memory consumption is to call the [see_memory_usage()](https://github.com/microsoft/DeepSpeed/blob/ff7d5275f2aa916cb5f320e0d817154e96f9cdb6/deepspeed/runtime/utils.py#L793) utility after model and optimizer initialization (e.g., `deepspeed.intialize()`). This utility leverages torch.cuda memory management [routines](https://pytorch.org/docs/stable/cuda.html#cuda-memory-management-api).

Hello, I have a question for ZeRO3 memory usage too. When I'm training a 13B model with 8 machines of 8 * A100, the maximum seq length can be set to 25k. However, when the machines num comes to 12, 25k seq length will lead to OOM. As you say, the activation seems to be a bottleneck here, but don't know why the seq length can't be the same.
